<br>
<font>
<div dir=ltr align=center>
<font color=0F5298 size=7>
    Introduction to Machine Learning <br>
<font color=2565AE size=5>
    Electrical Engineering Department <br>
    Spring 2024<br>
<font color=3C99D size=5>
    Practical Assignment 2 <br>
<font color=696880 size=4>

# Student Information


In [ ]:
Student_Name = "Radin Khayyam"
Student_ID = "99101579"

print("Student_Name:",Student_Name)
print("Student_ID:",Student_ID)

Student_Name: Radin Khayyam
Student_ID: 99101579


### **1- Data Preprocessing and Splitting:**

**1.1-**



In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

data = pd.read_csv('cardio_pred.csv')

print(data.head())
print('=================================================')
print(data.info())

    age  gender  height  weight  ap_hi  ap_lo  smoke  alco  active  cardio  \
0  51.0     1.0   168.0    62.0  110.0   80.0    0.0   0.0     1.0     0.0   
1  56.0     0.0   156.0    85.0  140.0   90.0    0.0   0.0     1.0     1.0   
2  52.0     0.0   165.0    64.0  130.0   70.0    0.0   0.0     0.0     1.0   
3  49.0     1.0   169.0    82.0  150.0    NaN    0.0   NaN     1.0     1.0   
4  49.0     0.0     NaN    56.0  100.0   60.0    0.0   0.0     0.0     0.0   

   cholesterol_1  cholesterol_2  cholesterol_3  gluc_1  gluc_2  gluc_3  \
0            1.0            0.0            0.0     1.0     0.0     0.0   
1            0.0            0.0            NaN     1.0     0.0     0.0   
2            0.0            0.0            1.0     1.0     0.0     NaN   
3            1.0            0.0            0.0     1.0     0.0     0.0   
4            1.0            0.0            NaN     1.0     0.0     0.0   

         bmi   bp  
0  21.967120  0.0  
1  34.927679  2.0  
2  23.507805  2.0  
3  28.

**1.2-**

In [ ]:
data_cleaned = data.dropna()

print(data_cleaned.head())
print('=================================================')
print(data_cleaned.info())

     age  gender  height  weight  ap_hi  ap_lo  smoke  alco  active  cardio  \
0   51.0     1.0   168.0    62.0  110.0   80.0    0.0   0.0     1.0     0.0   
6   62.0     0.0   157.0    93.0  130.0   80.0    0.0   0.0     1.0     0.0   
7   63.0     1.0   178.0    95.0  130.0   90.0    0.0   0.0     1.0     1.0   
12  41.0     1.0   165.0    60.0  120.0   80.0    0.0   0.0     0.0     0.0   
14  40.0     1.0   181.0    95.0  130.0   90.0    1.0   1.0     1.0     0.0   

    cholesterol_1  cholesterol_2  cholesterol_3  gluc_1  gluc_2  gluc_3  \
0             1.0            0.0            0.0     1.0     0.0     0.0   
6             0.0            0.0            1.0     1.0     0.0     0.0   
7             0.0            0.0            1.0     0.0     0.0     1.0   
12            1.0            0.0            0.0     1.0     0.0     0.0   
14            1.0            0.0            0.0     1.0     0.0     0.0   

          bmi   bp  
0   21.967120  0.0  
6   37.729725  2.0  
7   29.9835

**1.3-**

In [ ]:
train_set, test_set = train_test_split(data_cleaned, test_size=0.2, random_state=2024)

### **2- Logistic Regression:**

**2.1-**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

X_train = train_set.drop('cardio', axis=1)
y_train = train_set['cardio']
X_test = test_set.drop('cardio', axis=1)
y_test = test_set['cardio']

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy of the logistic regression model: {accuracy}')

Accuracy of the logistic regression model: 0.7448453608247423


**2.2-**

In [ ]:
coefficients = pd.DataFrame(model.coef_[0], index=X_train.columns, columns=['Coefficient'])
sorted_coefficients = coefficients.sort_values(by='Coefficient', ascending=False)
print(sorted_coefficients)

               Coefficient
cholesterol_3     0.568716
gluc_2            0.220131
bp                0.181182
weight            0.087546
age               0.056140
ap_hi             0.047104
ap_lo             0.013571
gluc_1           -0.060550
height           -0.064214
active           -0.108492
alco             -0.125764
smoke            -0.190576
bmi              -0.206846
gender           -0.213575
cholesterol_2    -0.242808
gluc_3           -0.250363
cholesterol_1    -0.416690




Positive coefficients increase the log-odds of the class label 1, meaning they increase the probability of having a cardiovascular disease. As we can see most important feature in cardiovascular disease is cholesterol_3.


### **3- SVM Classifier:**


**3.1-**

In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

C_values = [0.01, 0.1, 1, 10, 100]
print("Results for SVM with linear kernel:")
for C in C_values:
    model = SVC(kernel='linear', C=C)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"C = {C} -----------> Accuracy = {accuracy}")

Results for SVM with linear kernel:
C = 0.01 -----------> Accuracy = 0.7474226804123711
C = 0.1 -----------> Accuracy = 0.7448453608247423
C = 1 -----------> Accuracy = 0.7448453608247423
C = 10 -----------> Accuracy = 0.7371134020618557
C = 100 -----------> Accuracy = 0.7448453608247423


**3.2-**

In [ ]:

C_values = [0.01, 0.1, 1, 10, 100]
degrees = [2, 3, 4]
print("Results for SVM with poly kernel:")
for C in C_values:
  for degree in degrees:
      model = SVC(kernel='poly', degree=degree, C=C)
      model.fit(X_train, y_train)
      y_pred = model.predict(X_test)
      accuracy = accuracy_score(y_test, y_pred)
      print(f"C = {C}, Degree = {degree} -----------> Accuracy = {accuracy}")

Results for SVM with poly kernel:
C = 0.01, Degree = 2 -----------> Accuracy = 0.5077319587628866
C = 0.01, Degree = 3 -----------> Accuracy = 0.6855670103092784
C = 0.01, Degree = 4 -----------> Accuracy = 0.7010309278350515
C = 0.1, Degree = 2 -----------> Accuracy = 0.7216494845360825
C = 0.1, Degree = 3 -----------> Accuracy = 0.7268041237113402
C = 0.1, Degree = 4 -----------> Accuracy = 0.729381443298969
C = 1, Degree = 2 -----------> Accuracy = 0.7396907216494846
C = 1, Degree = 3 -----------> Accuracy = 0.7345360824742269
C = 1, Degree = 4 -----------> Accuracy = 0.7345360824742269
C = 10, Degree = 2 -----------> Accuracy = 0.7242268041237113
C = 10, Degree = 3 -----------> Accuracy = 0.7319587628865979
C = 10, Degree = 4 -----------> Accuracy = 0.7242268041237113
C = 100, Degree = 2 -----------> Accuracy = 0.7474226804123711
C = 100, Degree = 3 -----------> Accuracy = 0.75
C = 100, Degree = 4 -----------> Accuracy = 0.7319587628865979


**3.3-**

In [ ]:
C_values = [0.01, 0.1, 1, 10, 100]
degrees = [2, 3, 4]
gamma_values = ['scale', 'auto']
print("Results for SVM with rbf kernel:")
for C in C_values:
    for gamma in gamma_values:
        model = SVC(kernel='rbf', gamma=gamma, C=C)
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        accuracy = accuracy_score(y_test, y_pred)
        print(f"C = {C}, Gamma = {gamma}, Accuracy = {accuracy}")

Results for SVM with rbf kernel:
C = 0.01, Gamma = scale, Accuracy = 0.5051546391752577
C = 0.01, Gamma = auto, Accuracy = 0.5051546391752577
C = 0.1, Gamma = scale, Accuracy = 0.7242268041237113
C = 0.1, Gamma = auto, Accuracy = 0.5051546391752577
C = 1, Gamma = scale, Accuracy = 0.7371134020618557
C = 1, Gamma = auto, Accuracy = 0.6237113402061856
C = 10, Gamma = scale, Accuracy = 0.7242268041237113
C = 10, Gamma = auto, Accuracy = 0.6314432989690721
C = 100, Gamma = scale, Accuracy = 0.7396907216494846
C = 100, Gamma = auto, Accuracy = 0.6314432989690721


**3.4-**

The best result is for poly kernel with parameters C = 100 and degree = 3.
accuracy is 75% with these parameters. the reason is non linear relationship between inputs and the label.

### **4- Perceptron Classifier:**

**4.1-**

In [ ]:
import numpy as np

class SampleModePerceptron:
    def __init__(self, learning_rate=0.01, epochs=1000):
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.weights = None

    def fit(self, X, y):
        self.weights = np.zeros(X.shape[1] + 1)
        X_with_bias = np.c_[X, np.ones(X.shape[0])]

        for epoch in range(self.epochs):
            for xi, target in zip(X_with_bias, y):
                linear_output = np.dot(xi, self.weights)
                prediction = 1 if linear_output > 0 else 0
                self.weights += self.learning_rate * (target - prediction) * xi

    def predict(self, X):
        X_with_bias = np.c_[X, np.ones(X.shape[0])]
        linear_outputs = np.dot(X_with_bias, self.weights)
        predictions = np.where(linear_outputs > 0, 1, 0)
        return predictions

    def accuracy(self, X, y):
        predictions = self.predict(X)
        return np.mean(predictions == y)


perceptron = SampleModePerceptron(learning_rate=0.01, epochs=1000)
perceptron.fit(X_train.values, y_train.values)

test_accuracy = perceptron.accuracy(X_test.values, y_test.values)
print(f"Accuracy of the sample mode perceptron on the test set: {test_accuracy:.2f}")


Accuracy of the sample mode perceptron on the test set: 0.69


I tried both of them but the sample mode algorithm reached to a better accuracy.

**4.2-**

In [ ]:
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score

sk_perceptron = Perceptron(max_iter=100, eta0=0.01, random_state=2024)
sk_perceptron.fit(X_train, y_train)

sk_y_pred = sk_perceptron.predict(X_test)

sk_accuracy = accuracy_score(y_test, sk_y_pred)
print(f"Accuracy of sklearn's perceptron on the test set: {sk_accuracy:.2f}")


Accuracy of sklearn's perceptron on the test set: 0.61


In [ ]:
print(f"Custom Perceptron Accuracy: {test_accuracy:.2f}")
print(f"Sklearn Perceptron Accuracy: {sk_accuracy:.2f}")

if test_accuracy > sk_accuracy:
    print("Custom Perceptron performs better.")
elif test_accuracy < sk_accuracy:
    print("Sklearn Perceptron performs better.")
else:
    print("Both Perceptrons perform equally.")


Custom Perceptron Accuracy: 0.69
Sklearn Perceptron Accuracy: 0.61
Custom Perceptron performs better.


### **5- MLP Classifier:**

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

param_grid = {
    'mlpclassifier__hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 100)],
    'mlpclassifier__activation': ['tanh', 'relu'],
    'mlpclassifier__solver': ['sgd', 'adam'],
    'mlpclassifier__alpha': [0.0001, 0.001, 0.01],
    'mlpclassifier__max_iter': [500]
}


pipeline = make_pipeline(StandardScaler(), MLPClassifier(random_state=2024))

grid_search = GridSearchCV(pipeline, param_grid, cv=3, scoring='accuracy', n_jobs=-1)

grid_search.fit(X_train, y_train)

print("Best parameters:", grid_search.best_params_)
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

best_model = grid_search.best_estimator_
test_accuracy = best_model.score(X_test, y_test)
print(f"Test set accuracy: {test_accuracy:.2f}")


Best parameters: {'mlpclassifier__activation': 'relu', 'mlpclassifier__alpha': 0.01, 'mlpclassifier__hidden_layer_sizes': (100, 100), 'mlpclassifier__max_iter': 500, 'mlpclassifier__solver': 'sgd'}
Best cross-validation score: 0.72
Test set accuracy: 0.74


### **6- Decision Tree:**

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 10, 20],
    'min_samples_leaf': [1, 5, 10],
    'criterion': ['gini', 'entropy']
}

dtree = DecisionTreeClassifier(random_state=2024)

grid_search = GridSearchCV(dtree, param_grid, cv=5, scoring='accuracy', n_jobs=-1)

grid_search.fit(X_train, y_train)

print("Best parameters:", grid_search.best_params_)
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

best_tree = grid_search.best_estimator_
test_accuracy = best_tree.score(X_test, y_test)
print(f"Test set accuracy: {test_accuracy:.2f}")


Best parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 10, 'min_samples_split': 2}
Best cross-validation score: 0.68
Test set accuracy: 0.69


### **7- Compare Results and Discussion:**

Best accuracy is 74% which is for MLP classifer with two hidden layers with 100 neurons, relu activation function and 500 iteration and using sgd solver.
In addition to accuracy we can calculate othe measures like sensitivity, specificity, precision and F1 score to evaluate the performance of the model.